snuba/pipeline/query_pipeline.py

```diff
@@ -1,44 +1,79 @@
+    a logical single entity query and request settings, Performs the
+    query processing steps to the point where the query is ready to
+    be executed.
+
+    the query is returned as a ClickhouseQueryPlan.
+
+    TODO: If we will build a processing pipeline for composite query
+    the return type of this pipeline will have to change.
+    """
+
+    @abstractmethod
+    def execute(self, input: EntityProcessingPayload) -> ClickhouseQueryPlan:
+        raise NotImplementedError
 
-class QueryPipeline(ABC):
+
+class QueryExecutionPipeline(Segment[Request, QueryResult], ABC):
     """
-    Contains the instructions to build the query plans for the requested query.
-    The QueryPipeline runs the query plan processors, executes the query plans and
-    returns the relevant result.
+    Contains the instructions to execute a query.
+    The QueryExecutionPipeline performs the all query processing steps and,
+    executes the query plan and returns the result.
 
-    Most of the time, a single query plan is built by the SingleQueryPlanPipeline,
+    Most of the time, a single query plan is built by the SinglePlanExecutionPipeline,
     however under certain circumstances we may have a pipeline with multiple query
     plans. For example, the MultipleQueryPlanPipeline provides a way to build and
     execute more than one query plan and compare their results, which provides a
     way to experiment with different query plans in production without actually
     using their results yet.
 
     This component is produced by the QueryPipelineBuilder.
-
-    TODO: In future, the query pipeline wiil be composed of segments, which will
-    each perform individual processing steps and can be easily reused.
-    Examples of transformations performed by segments are entity processing, query
-    plan building and processing, storage processing, as well as splitters
-    and collectors to help break apart and reassemble joined queries.
     """
 
     @abstractmethod
-    def execute(self) -> QueryResult:
+    def execute(self, input: Request) -> QueryResult:
         raise NotImplementedError
 
 
 class QueryPipelineBuilder(ABC):
     """
-    Builds a query pipeline, which contains the directions for building and running
-    query plans.
+    Builds a query pipeline, which contains the directions for building
+    and running query plans.
     """
 
     @abstractmethod
-    def build_pipeline(self, request: Request, runner: QueryRunner) -> QueryPipeline:
+    def build_execution_pipeline(
+        self, request: Request, runner: QueryRunner
+    ) -> QueryExecutionPipeline:
+        """
+        Returns a pipeline to execute a query
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def build_processing_pipeline(self, request: Request) -> QueryProcessingPipeline:
+        """
+        Returns a pipeline that executes the processing phase of a query.
+
+        TODO: Split this method out of this class so that we can reuse
```

Comment:
I am not 100% sure yet we will need the dataset to return the processing pipeline for the composite query. That will also depend on how we integrate the pipeline delegator (which will be decided tomorrow-ish). If the composite pipeline will be static, likely there will be no need for a pipeline builder at that level.

Reply:
not relevant anymore